plugin/storage/es/samplingstore/storage.go

```diff
@@ -0,0 +1,214 @@
+	ctx := context.Background()
+	clientFn := s.client()
+	indices, err := getLatestIndices(s.samplingIndexPrefix, s.indexDateLayout, clientFn, s.samplingIndexRolloverFrequency, s.maxSampleTime)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get latest indices: %w", err)
+	}
+	searchResult, err := clientFn.Search(indices...).
+		Size(s.maxDocCount).
+		IgnoreUnavailable(true).
+		Do(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to search for Latest Probabilities: %w", err)
+	}
+	lengthOfSearchResult := len(searchResult.Hits.Hits)
+	if lengthOfSearchResult == 0 {
+		return nil, nil
+	}
+
+	var latestProbabilities dbmodel.TimeProbabilitiesAndQPS
+	latestTime := time.Time{}
+	for _, hit := range searchResult.Hits.Hits {
+		var data dbmodel.TimeProbabilitiesAndQPS
+		if err = json.Unmarshal(*hit.Source, &data); err != nil {
+			return nil, fmt.Errorf("unmarshalling documents failed: %w", err)
+		}
+		if data.Timestamp.After(latestTime) {
+			latestTime = data.Timestamp
+			latestProbabilities = data
+		}
+	}
+	return latestProbabilities.ProbabilitiesAndQPS.Probabilities, nil
+}
+
+func (s *SamplingStore) writeThroughput(indexName string, ts time.Time, throughputs []*model.Throughput) {
+	s.client().Index().Index(indexName).Type(throughputType).
+		BodyJson(&dbmodel.TimeThroughput{
+			Timestamp:  ts,
+			Throughput: dbmodel.FromThroughputs(throughputs),
+		}).Add()
+}
+
+func (s *SamplingStore) writeProbabilitiesAndQPS(indexName string, ts time.Time, pandqps dbmodel.ProbabilitiesAndQPS) {
+	s.client().Index().Index(indexName).Type(probabilitiesType).
+		BodyJson(&dbmodel.TimeProbabilitiesAndQPS{
+			Timestamp:           ts,
+			ProbabilitiesAndQPS: pandqps,
+		}).Add()
+}
+
+func getLatestIndices(indexPrefix, indexDateLayout string, clientFn es.Client, reduceDuration time.Duration, maxDuration time.Duration) ([]string, error) {
+	ctx := context.Background()
+	now := time.Now().UTC()
+	end := now.Add(-maxDuration)
+	for {
+		indexName := indexWithDate(indexPrefix, indexDateLayout, now)
+		exists, err := clientFn.IndexExists(indexName).Do(ctx)
+		if err != nil {
+			return nil, err
+		}
+		if exists {
+			return []string{indexName}, nil
+		}
+		if now == end {
+			return nil, fmt.Errorf("falied to find latest index")
+		}
+		now = now.Add(reduceDuration)
+	}
+}
+
+func getReadIndices(indexName, indexDateLayout string, startTime time.Time, endTime time.Time, reduceDuration time.Duration) []string {
+	var indices []string
+	firstIndex := indexWithDate(indexName, indexDateLayout, startTime)
+	currentIndex := indexWithDate(indexName, indexDateLayout, endTime)
+	for currentIndex != firstIndex {
+		if len(indices) == 0 || indices[len(indices)-1] != currentIndex {
+			indices = append(indices, currentIndex)
+		}
+		endTime = endTime.Add(reduceDuration)
+		currentIndex = indexWithDate(indexName, indexDateLayout, endTime)
+	}
+	indices = append(indices, firstIndex)
```

Comment:
the logic above is rather convoluted, could we not make it simpler?
```suggestion
	lastIndex := indexWithDate(indexName, indexDateLayout, endTime)
	indices := []string{lastIndex} // always include the latest
	currentIndex := indexWithDate(indexName, indexDateLayout, startTime)
	for currentIndex != lastIndex {
		indices = append(indices, currentIndex)
		startTime = startTime.Add(lookback)
		currentIndex = indexWithDate(indexName, indexDateLayout, startTime)
	}
```

Reply:
I am wondering if the order or position of indices matters because the tests are currently failing due to it.
In suggested logic : if start=10 feb, end = 12 feb, we get : 
```go
"prefix-jaeger-sampling-2024-02-12",
"prefix-jaeger-sampling-2024-02-10", 
"prefix-jaeger-sampling-2024-02-11"
```
if order matters, then this code lgtm : 
```go
var indices []string
firstIndex := indexWithDate(indexName, indexDateLayout, startTime)
currentIndex := indexWithDate(indexName, indexDateLayout, endTime)
for currentIndex != firstIndex {
	indices = append(indices, currentIndex)
	endTime = endTime.Add(lookback)
	currentIndex = indexWithDate(indexName, indexDateLayout, endTime)
}
indices = append(indices, firstIndex)
```

Reply:
whether the order matters or not depends on how you use the returned data. Since you're picking the latest entry based on the timestamp, the order should not matter. However, in theory it's more efficient to search indices in reverse chronological order because the most recent one is most likely to have the needed data. Instead, if you search through all indices in the lookback interval then you will be crunching a lot of data on every iteration, even though you're only interested in the latest entry.

Reply:
ok thanks, got it.